Vmstat Comparision with and withould load.

DeepthiYV opened the pull request:
- Related ticket: https://progress.opensuse.org/issues/48959
- Needles: NO
- Verification run: 
SLE15 SP2 - http://10.161.229.147/tests/287#step/vmstat/1
SLE15 SP1 - http://10.161.229.147/tests/254#step/vmstat/1
Tumbelweed- http://10.161.229.147/tests/287#step/vmstat/1

punkioudi reviewed (commented):
Small correction in the title, it should be `without` ^^

dzedro commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,91 @@
+    my $min_cpu_lightload = min(@cpu_lite_load);
+
+    # Increasing the load in the system using package stress-ng.
+    add_suseconnect_product(get_addon_fullname('phub')) if (is_sle);
```
Remove this line, it;s in the mau-extratestss-hub because there is phub added at beginning.
Also schedule it only there remove it from `schedule/functional/extra_tests_textmode.yaml`

dzedro commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,91 @@
+# SUSE's openQA tests
+#
+# Copyright © 2019 SUSE LLC
```
2021

punkioudi commented:
Schedule in TW seems to be missing

okurz reviewed (changes requested):
In the PR title:
* no dot at EOL
* s/Comparision/Comparison/

The commit message could have a bit more details, e.g. what issue is fixed, why this needs to change, to which versions of which product it applies, link to a bug or a fate entry, the choices you made, etc. Also see https://commit.style/ or http://chris.beams.io/posts/git-commit/ as a helpful guide how to write good commit messages.

okurz commented on lib/console/vmstat_utils.pm:
```diff
@@ -0,0 +1,69 @@
+
+    my $file_name = shift;
+    my @res_mem   = split '\n', script_output "awk '{print \$4}' $file_name";
+    my (@mem_num, @cpu_num);
+    foreach my $elem (@res_mem) {
+        if ($elem =~ m/\d+/) {
+            push @mem_num, $elem;
+        }
+    }
```
I suggest to use map or grep instead. Something like

```suggestion
    my @mem_num = grep { m/\d+/ } @res_mem;
```

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,91 @@
+    my $avg_mem_heavyload = average(@mem_hvy);
+    my $avg_cpu_heavyload = average(@cpu_hvy);
+
+    # Comparision of the vmstat with and without load
```
s/Comparision/Comparison/

foursixnine commented on schedule/functional/extra_tests_textmode.yaml:
```diff
@@ -141,6 +141,7 @@ schedule:
     - console/rpcbind
     - console/timezone
     - console/procps
+    - console/vmstat
```
This change belongs to schedule/functional/extra_tests_textmode_phub.yaml

foursixnine commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,91 @@
+    my $avg_cpu_heavyload = average(@cpu_hvy);
+
+    # Comparision of the vmstat with and without load
+    if ($avg_mem_heavyload > $min_mem_lightload or $avg_cpu_heavyload < $min_cpu_lightload) {
+        die "Error!! vmstat sample output during light/heavy load";
+    }
+}
```
Make sure that you show the expected and obtained values, to help debugging

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,91 @@
+
+    # Comparision of the vmstat with and without load
+    if ($avg_mem_heavyload > $min_mem_lightload or $avg_cpu_heavyload < $min_cpu_lightload) {
+        die "Error!! vmstat sample output during light/heavy load";
```
please don't, see https://wiki.lspace.org/mediawiki/Multiple_exclamation_marks :)

okurz commented on lib/console/vmstat_utils.pm:
```diff
@@ -0,0 +1,62 @@
+# Summary: Basic functions for testing vmstat.
+# Maintainer: QE Core <qe-core@suse.de>
+
+package console::vmstat_utils;
```
I wonder, why do you separate out this file and not keep the implementation in the only test module file that currently uses the functionality? Commonly lib/ should be used for functionality that is used in more than one test module

DeepthiYV replied:
This was a suggestion by other colleague to keep only main function in test module file and remaining functionality in separate file commonly /lib.

okurz replied:
ok, can you explain why? Sorry, I did not see that suggestion

DeepthiYV replied:
It was during pair programming session.

okurz commented on lib/console/vmstat_utils.pm:
```diff
@@ -0,0 +1,62 @@
+
+    my $file_name = shift;
+    my @res_mem   = split '\n', script_output "awk '{print \$4}' $file_name";
+    my (@mem_num, @cpu_num);
+    @mem_num = map { $_ =~ m/\d+/ } @res_mem;
+    my @res_cpu = split '\n', script_output "awk '{print \$13}' $file_name";
+    @cpu_num = map { $_ =~ m/\d+/ } @res_cpu;
```
```suggestion
    my @mem_num = map { $_ =~ m/\d+/ } @res_mem;
    my @res_cpu = split '\n', script_output "awk '{print \$13}' $file_name";
    my @cpu_num = map { $_ =~ m/\d+/ } @res_cpu;
```

DeepthiYV replied:
Have done changes in code.

okurz commented on lib/console/vmstat_utils.pm:
```diff
@@ -0,0 +1,62 @@
+ Returns the average of numbers.
+
+=cut
+sub average {
+    my @Array_To_Average = shift;
+    my $size             = @Array_To_Average;
+    my $total            = 0;
+    for (@Array_To_Average)
+    {
+        $total += $_;
+    }
+
+    return $total / $size;
+}
```
```suggestion
use List::Utils qw(sum);
…
sub average {
    my (@a) = @_;
    my $size = @a;
    return sum(@a) / $size;
}
```

but are you sure you want integer division?

DeepthiYV replied:
yes, further am comparing this average number with minimum of light load.

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,92 @@
+    $vm = 50;
+    my $stressng_pkg = "nohup stress-ng --cpu $cpu --vm $vm --vm-bytes 256m --timeout 60s&";
+    enter_cmd "$stressng_pkg";
+    sleep 10;
```
You are using `sleep()` or are simulating a sleep-like functionality. This is introducing the risk of either unnecessarily wasting time or just shifting race conditions and making them even harder to investigate. You *can* use sleep but there are only very limited cases where is the best solution (e.g. yield threads in low-level system code). Here you have the classical problem in software design of [Synchronization](https://en.wikipedia.org/wiki/Synchronization_(computer_science)) . sleep should not be used in place of proper synchronization methods. https://blogs.msmvps.com/peterritchie/2007/04/26/thread-sleep-is-a-sign-of-a-poorly-designed-program/ is one of multiple references that explains this: "The original problem is likely a timing/synchronization issue, ignoring it by hiding it with Thread.Sleep is only going to delay the problem and make it occur in random, hard to reproduce ways." And if the problem does not occur at least time is wasted. https://testing.googleblog.com/2021/03/test-flakiness-one-of-main-challenges.html states that as well: "Do NOT add arbitrary delays as these can become flaky again over time and slow down the test unnecessarily."

We are waiting because some condition changes some time ... keyword(s) is/are *some time*. It is important to know the exact condition that the code should wait for and check for that instead of delaying program execution for an arbitrary amount of time. For openQA tests in general just try to do what you as a human user would also do. You would not look on your watch waiting for exactly 100s before you look at the screen again, right? :)

See https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/CONTRIBUTING.md#coding-style for more details

DeepthiYV replied:
I am collecting vmstat output after the 10s delay of  stress-ng run, which is running in the background and so kept sleep of 10s. Later I will compare the vmstat sample with light load.

okurz replied:
Well, as I stated, you are waiting for *something*, so wait for that, not sleep 10

foursixnine replied:
Let's keep the sleep for the time being, the extra work needed to check whether the load did really increase due to stress-ng will be catched up anyway further ahead.

DeepthiYV replied:
we will keep the sleep.

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,92 @@
+use strict;
+use warnings;
+use POSIX;
+use version_utils;
```
I don't think you use that anymore, do you?

DeepthiYV replied:
I have removed "version_utils " which I dont use.

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,92 @@
+use console::vmstat_utils;
+
+sub run {
+
```
please delete the blank line

DeepthiYV replied:
I have deleted :)

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,92 @@
+    my ($vm, $cpu);
+    $self->select_serial_terminal;
+    zypper_call('in procps');
+    assert_script_run("rpm -q procps");
```
is this just for debugging?

DeepthiYV replied:
yes, after installation just displaying installed procps version.

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,92 @@
+    validate_script_output("vmstat",
+        qr/(procs\s-+memory-+\s-+swap-+\s-+io-+\s-+system-+\s-+cpu-+).*(\s+|\d+)+/s);
+
+    # Extend vmstat utility. Verifying the vmstat with options.
```
there is something missing in the sentence "Verifying the …"

DeepthiYV replied:
I have added missing the sentence.     # Extend vmstat utility. Verifying the vmstat report information with options and comparing the output with and without load.

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,92 @@
+    $self->select_serial_terminal;
+    zypper_call('in procps');
+    assert_script_run("rpm -q procps");
+    validate_script_output("vmstat",
```
I suggest by default to always use single tick strings unless you want to evaluate perl variables within strings

DeepthiYV replied:
Thank you for the suggestion and I will leave it like this.

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,92 @@
+        qr/(procs\s-+memory-+\s-+swap-+\s-+io-+\s-+system-+\s-+cpu-+).*(\s+.*\s+inact\s+active\s*.*).*(\s+|\d+)+/s);
+    validate_script_output("vmstat -f", qr/(\d+\s+forks)/s);
+
+    #Running vmstat under light load and dumping the output to a file. Calculate and collect the minimum idle memory from the output.
```
Please add a space after `#`

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,92 @@
+    assert_script_run("rpm -q stress-ng");
+    my $total_cpu = script_output("cat /proc/cpuinfo | grep processor | wc -l");
+    my $total_mem = script_output("free -h | awk '/Mem\:/ { print \$2 }'");
+    if ($total_cpu == 1) { $cpu = $total_cpu; } else { $cpu = floor($total_cpu * 0.75); }
```
please delete the semicolons within the simple statements within if/else bodies

DeepthiYV replied:
Delete the semicolons :)

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,92 @@
+    assert_script_run("cat /tmp/output_load.log");
+
+    #kill the running background process stress-ng.
+    assert_script_run("pkill -9 stress-ng");
```
please avoid `-9` unless there is a good reason why you need it. It's bad habit to resort to killing instead of terminating unless there is a good reason for it.

DeepthiYV replied:
Okay. I have modified in the code. Thank you .
  # kill the running background process stress-ng.
   assert_script_run("killall stress-ng");

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,92 @@
+    if ($avg_mem_heavyload > $min_mem_lightload or $avg_cpu_heavyload < $min_cpu_lightload) {
+        my $error_msg = "wrong vmstat information after the heavy load. Expected $avg_mem_heavyload must be less than $min_mem_lightload and $avg_cpu_heavyload must be greaterthan $min_cpu_lightload";
+        die($error_msg);
+
+    }
+
+}
+
+1;
+
```
looks like you like blank lines too much :D

okurz commented on tests/console/vmstat.pm:
```diff
@@ -0,0 +1,92 @@
+
+    # Comparison  of the vmstat with and without load
+    if ($avg_mem_heavyload > $min_mem_lightload or $avg_cpu_heavyload < $min_cpu_lightload) {
+        my $error_msg = "wrong vmstat information after the heavy load. Expected $avg_mem_heavyload must be less than $min_mem_lightload and $avg_cpu_heavyload must be greaterthan $min_cpu_lightload";
```
please wrap consistently at 160 characters, I think this is longer, see https://github.com/os-autoinst/os-autoinst/blob/master/.perltidyrc#L3

foursixnine replied:
I tested the -l parameter of tidy (changed it to 72), it doesn't seem to work (on a birds eye view)

DeepthiYV replied:
Updated the error_msg.

foursixnine reviewed (approved):
@DeepthiYV can you provide a verification run with the latest code?

punkioudi reviewed (changes requested):
It still seems that the the test is not scheduled for TW

DeepthiYV commented:
> It still seems that the test is not scheduled for TW

scheduled for tumbelweed as well. I have added execution link as well.

punkioudi commented:
> > It still seems that the test is not scheduled for TW
> 
> scheduled for tumbleweed as well. I have added execution link as well.

where? I have seen the VR for TW, but I don't see the scheduling for TW :)

DeepthiYV commented:
> > > It still seems that the test is not scheduled for TW
> > 
> > 
> > scheduled for tumbleweed as well. I have added execution link as well.
> 
> where? I have seen the VR for TW, but I don't see the scheduling for TW :)

I have added vmstat test in the file  schedule/functional/extra_tests_textmode.yaml  in opensuse_tests, which is scheduler for TW. Hope I have understood the scheduling part. Let me know if I had missed anything.

punkioudi reviewed (approved):
lgtm!

foursixnine commented:
@okurz any other comments before merging?

okurz reviewed (changes requested):
well, I don't see my suggestions implemented. Of course, they are only suggestions so they don't need to be followed. I suggest to explicitly respond to each review comment, either if the suggestion has been followed and implemented or even if one does not want to follow a suggestion, but then just tell the reviewer. This way we can ensure that no comment was overlooked.

dzedro commented on schedule/functional/extra_tests_textmode.yaml:
```diff
@@ -66,6 +66,7 @@ conditional_schedule:
                 - '{{steamcmd}}'
                 - console/libqca2
                 - console/python_scientific
+                - console/vmstat
```
C&P
Remove this line, it's in the mau-extratests-phub because there is phub added at beginning.
Also schedule it only there, remove it from schedule/functional/extra_tests_textmode.yaml

DeepthiYV replied:
This is scheduled for Tumbleweed and Leap.

dzedro replied:
But not only there ... https://openqa.suse.de/tests/5975595

dzedro replied:
My bad it's ok, it's opnsuse part ...

foursixnine commented:
Ok let's see what else breaks :)
